packages/workspace-resolver/src/resolve-dependencies/index.ts

```diff
@@ -0,0 +1,145 @@
+import type { WorkspaceDependencyObject } from '@modular-scripts/modular-types';
+
+type OrderedDependencies = Map<string, number>;
+type OrderedUnvisited = { name: string; level: number };
+type OptionalWorkspaceDependencyObject = Partial<WorkspaceDependencyObject>;
+
+export function computeDescendantSet(
+  workspaceNames: string[],
+  allWorkspaces: Record<string, OptionalWorkspaceDependencyObject>,
+): Set<string> {
+  // Really simple and performant ancestors collection: walk the graph via BFS and collect all the dependencies encountered in a set.
+  // If one dependency was already encountered before, don't process it. This is cycle resistant.
+  const unvisited: string[] = [...workspaceNames];
+  const visited: Set<string> = new Set();
+
+  while (unvisited.length) {
+    const currentDependency = unvisited.shift();
+    if (!currentDependency) break;
+    visited.add(currentDependency);
+
+    const immediateDependencies =
+      allWorkspaces[currentDependency]?.workspaceDependencies;
+
+    if (immediateDependencies) {
+      for (const immediateDep of immediateDependencies) {
+        if (!visited.has(immediateDep)) {
+          unvisited.push(immediateDep);
+        }
+      }
+    }
+  }
+  return setDiff(visited, new Set(workspaceNames));
+}
+
+export function computeAncestorSet(
+  workspaceNames: string[],
+  allWorkspaces: Record<string, WorkspaceDependencyObject>,
+): Set<string> {
+  // Computing an ancestor set is like computing a dependant set with an inverted graph
+  return computeDescendantSet(
+    workspaceNames,
+    invertDependencyDirection(allWorkspaces),
+  );
+}
+
+// This function takes a tree of dependencies (dependant -> child dependencies)
+// and returns an equivalent tree where the relation's direction is inverted
+// (dependency -> parent dependencies)
+// This allows us to use the same algorithm to query ancestors or descendants.
+export function invertDependencyDirection(
```

Comment:
Does `invertDependencyDirection` currently handle cycles? I'm wondering if it's currently assuming acyclic graphs.

I guess if not, it should throw, like how `traverseWorkspaceRelations` does

Reply:
It handles cycles, as it just inverts the tree representation of a graph into another tree representation.

Reply:
Fair enough 👍